lib/puppet/resource.rb

```diff
@@ -431,24 +431,23 @@ def key_attributes
     resource_type.respond_to?(:key_attributes) ? resource_type.key_attributes : [:name]
   end
 
-  # Convert our resource to yaml for Hiera purposes.
-  def to_hierayaml
-    # Collect list of attributes to align => and move ensure first
-    attr = parameters.keys
-    attr_max = attr.inject(0) { |max,k| k.to_s.length > max ? k.to_s.length : max }
+  # Convert our resource to a hash suitable for YAML serialization.
+  def to_hiera_hash
+    res = {}
 
-    attr.sort!
-    if attr.first != :ensure  && attr.include?(:ensure)
-      attr.delete(:ensure)
-      attr.unshift(:ensure)
+    if parameters.include?(:ensure)
+      res['ensure'] = parameters.delete(:ensure).to_s
+    end
+    parameters.sort.each do |k,v|
+      res[k.to_s] = case v
+                    when Time, Symbol
+                      v.to_s
+                    else
+                      v
+                    end
     end
```

Comment:
I suppose we will want to do this conversion to string for nested structures members as well.